x-pack/osquerybeat/beater/osquerybeat.go

```diff
@@ -49,6 +51,9 @@ type osquerybeat struct {
 	// Beat lifecycle context, cancelled on Stop
 	cancel context.CancelFunc
 	mx     sync.Mutex
+
+	// limiter to run one query at a time
+	queryLimiterCh chan struct{}
```

Comment:
The way it is initialized and used looks like a counting semaphore. 

I think Heartbeat uses this type: https://pkg.go.dev/golang.org/x/sync/semaphore

Reply:
Checked the implementation of that, it's more than I need and didn't like this property:
"If ctx is already done, Acquire may still succeed without blocking."
https://pkg.go.dev/golang.org/x/sync/semaphore#Weighted.Acquire

Reply:
This property is only true if all these conditions are met:
- the current go-routine is waiting on acquire
- The ctx is closed at the same time as acquire succeeds
- Another go-routine did actually call release when the context was closed.

As the `select` clause randomized checking order, your code actually has the same property, but worse. The Acquire fails if the ctx was closed before. Your select based approach can succeed even if the context was already closed.

Reply:
Yep, the query below will fail with cancelled context. Will update to sem.